File: src/main/scala/actor/ArtistWriterActor.scala
package actor

import akka.actor.Actor
import akka.pattern.ask
import akka.util.Timeout
import base._
import com.mongodb.casbah.Imports._
import concurrent.Future
import file.{Location, Artist}


class ArtistWriterActor extends Actor {

  implicit val timeout = Timeout(10000)

  import context.dispatcher

  private val progressListener = context.actorFor("akka://LoadingSystem/user/progressListener")

  private val artistsCollection = context.actorFor("akka://LoadingSystem/user/artists")
  private val locationsCollection = context.actorFor("akka://LoadingSystem/user/locations")
  private val similarititesCollection = context.actorFor("akka://LoadingSystem/user/similaritites")
  private val songsCollection = context.actorFor("akka://LoadingSystem/user/songs")
  private val tagsCollection = context.actorFor("akka://LoadingSystem/user/tags")
  private val termsCollection = context.actorFor("akka://LoadingSystem/user/terms")

  def receive = {
    case LoadArtist(source) => {
      val artist: Artist = Artist(source)

      val locations = locationOf(artist)
      val songs = songsOf(artist)
      val similars = similarsOf(artist)
      val tags = tagsOf(artist)
      val terms = termsOf(artist)

      val details: Future[DBObject] = for {
        location <- locations
        songs <- songs
        similarities <- similars
        tags <- tags
        terms <- terms
      }
      yield {
        val detailBuilder = MongoDBObject.newBuilder

        location.map(detailBuilder += "location" -> _)

        detailBuilder += "songs" -> songs.toList

        detailBuilder += "similarities" -> similarities.toList

        detailBuilder += "mbtags" -> tags.toList

        detailBuilder += "terms" -> terms.toList

        detailBuilder.result()
      }

      details.map(details => artistsCollection.tell(Write(List(artist.toMongo ++ details)), sender = progressListener))

    }
  }

  private def locationOf(artist: Artist): Future[Option[BasicDBObject]] = {
    ask(locationsCollection, FindOne(Location.byArtistName(artist.name))).mapTo[Option[BasicDBObject]]
  }

  private def songsOf(artist: Artist): Future[MongoCursor] = {
    ask(songsCollection, Find(Song.byArtistName(artist.name))).mapTo[MongoCursor]
  }

  private def similarsOf(artist: Artist): Future[MongoCursor] = {
    ask(similarititesCollection, Find(ArtistSimilarity.byTarget(artist.name))).mapTo[MongoCursor]
  }

  private def tagsOf(artist: Artist): Future[MongoCursor] = {
    ask(tagsCollection, Find(TagByArtist.byArtistId(artist.id))).mapTo[MongoCursor]
  }

  private def termsOf(artist: Artist): Future[MongoCursor] = {
    ask(termsCollection, Find(TermByArtist.byArtistId(artist.id))).mapTo[MongoCursor]
  }
}

File: src/main/java/actor/ArtistWriterActor.java
package actor;

import actor.message.*;
import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import akka.dispatch.Future;
import akka.dispatch.Mapper;
import akka.dispatch.OnComplete;
import akka.util.Timeout;
import base.ArtistSimilarity;
import base.Song;
import base.TagByArtist;
import base.TermByArtist;
import file.Artist;
import file.Location;

import static akka.pattern.Patterns.ask;
import static file.Artist.toArtist;

public class ArtistWriterActor extends UntypedActor {

    private static final Timeout timeout = Timeout.apply(10000);

    private ActorRef progressListener = getContext().actorFor("akka://LoadingSystem/user/progressListener");

    private ActorRef artistsCollection = getContext().actorFor("akka://LoadingSystem/user/artists");
    private ActorRef locationsCollection = getContext().actorFor("akka://LoadingSystem/user/locations");
    private ActorRef similarititesCollection = getContext().actorFor("akka://LoadingSystem/user/similaritites");
    private ActorRef songsCollection = getContext().actorFor("akka://LoadingSystem/user/songs");
    private ActorRef tagsCollection = getContext().actorFor("akka://LoadingSystem/user/tags");
    private ActorRef termsCollection = getContext().actorFor("akka://LoadingSystem/user/terms");

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof LoadArtist) {
            Artist artist = toArtist(((LoadArtist) message).source);

            /*Future<Location> location = locationOf(artist);
            Future<Iterable<Song>> songs = songsOf(artist);
            Future<Iterable<ArtistSimilarity>> similars = similarsOf(artist);
            Future<Iterable<TagByArtist>> tags = tagsOf(artist);
            Future<Iterable<TermByArtist>> terms = termsOf(artist);*/

            Future<Object> status = ask(artistsCollection, new Write(new Artist[]{artist}), timeout);

            status.onComplete(toProgressListener(artist.getName()));


        } else {
            unhandled(message);
        }
    }

    private Future<Location> locationOf(Artist artist) {
        return ask(locationsCollection, new FindOne<Location>(String.format("{artistName: %s}", artist.getName()), Location.class), timeout)
                .map(new Mapper<Object, Location>() {
                    @Override
                    public Location apply(Object input) {
                        return (Location) input;
                    }
                });
    }

    private Future<Iterable<Song>> songsOf(Artist artist) {
        return ask(songsCollection, new Find<Song>(String.format("{artistName: %s}", artist.getName()), Song.class), timeout)
                .map(new Mapper<Object, Iterable<Song>>() {
                    @Override
                    public Iterable<Song> apply(Object input) {
                        return (Iterable<Song>) input;
                    }
                });
    }

    private Future<Iterable<ArtistSimilarity>> similarsOf(Artist artist) {
        return ask(similarititesCollection, new Find<ArtistSimilarity>(String.format("{target : %s}", artist.getName()), ArtistSimilarity.class), timeout)
                .map(new Mapper<Object, Iterable<ArtistSimilarity>>() {
                    @Override
                    public Iterable<ArtistSimilarity> apply(Object input) {
                        return (Iterable<ArtistSimilarity>) input;
                    }
                });
    }

    private Future<Iterable<TagByArtist>> tagsOf(Artist artist) {
        return ask(tagsCollection, new Find<TagByArtist>(String.format("{artist_id: %s}", artist.getId()), TagByArtist.class), timeout)
                .map(new Mapper<Object, Iterable<TagByArtist>>() {
                    @Override
                    public Iterable<TagByArtist> apply(Object input) {
                        return (Iterable<TagByArtist>) input;
                    }
                });
    }

    private Future<Iterable<TermByArtist>> termsOf(Artist artist) {
        return ask(termsCollection, new Find<TermByArtist>(String.format("{artist_id: %s}", artist.getId()), TermByArtist.class), timeout)
                .map(new Mapper<Object, Iterable<TermByArtist>>() {
                    @Override
                    public Iterable<TermByArtist> apply(Object input) {
                        return (Iterable<TermByArtist>) input;
                    }
                });
    }

    private OnComplete<Object> toProgressListener(final String successfulMessage) {
        return new OnComplete<Object>() {
            @Override
            public void onComplete(Throwable failure, Object success) throws Throwable {
                if (failure != null) {
                    progressListener.tell(failure);
                } else {
                    progressListener.tell(new Done(successfulMessage));
                }
            }
        };
    }
}
